[SOCRadar] Update legacy code

Powlinett opened the pull request:
### Proposed changes

- Update config:
  * remove 'radar_' suffix of config vars (from _config.yml_) 
  * add runtime validation (with `pydantic_settings` lib)
  * deprecate `RADAR_RUN_INTERVAL` (replaced by `CONNECTOR_DURATION_PERIOD`)
  * deprecate `RADAR_COLLECTIONS_UUID` (replace by `RADAR_FEED_LISTS` - flat version)
  * update _config.yml.sample_ / add _.env.sample_
  * update README
- Update Docker files:
  * add files from templates
  * add env vars examples in _docker-compose.yml_
  * update _Dockerfile_
- Replace deprecated code:
  * use `self.helper.schedule_process` method
  * use `self.helper.connector_logger` log methods
  * use `self.helper.create_stix2_bundle` method
- Error handling improvement:
  * add traceback in _main.py_
  * add 2 abstraction layers: `RadarAPIClient` and `ConverterToStix`
  * wrap expected errors with `RadarAPIError` and `ConverterError`
  * handle expected errors in application layer
  * close work gracefully in case of unexpected error

### Related issues

* #4193 
* #4194 

### Checklist

- [ ] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### Further comments

- Conversion to STIX refactor:
  * despite refactor, do _not_ change any mapping, skipping, or ingesting behavior
- Config changes:
  * In order to add config runtime validation, config vars had to follow same format as env vars, meaning that `radar_` suffix has been removed from `radar` config vars.

    _config.yml_ **before**
    ```yaml
    opencti:
      ...
    connector:
      ...
    radar:
      radar_base_feed_url: "http://example.com"
      radar_socradar_key: "api_key"
    ```
    _config.yml_ **now**
    ```yaml
    opencti:
      ...
    connector:
      ...
    radar:
      base_feed_url: "http://example.com"
      socradar_key: "api_key"
    ```
    It creates **breaking changes in development mode only**. cc @romain-filigran 
  * For sake of simplicity, I introduced a new env/config var, `RADAR_FEED_LISTS`, which is a flat version of `RADAR_COLLECTIONS_UUID` (now deprecated - but still handled).
    
    _config.yml_ **before**
    ```yaml
    opencti:
      ...
    connector:
      ...
    radar:
      radar_collections_uuid:
        collection_1:
          id: ["COLLECTION_UUID"]
          name: ["COLLECTION_NAME"]
        collection_2:
          id: ["COLLECTION_UUID"]
          name: ["COLLECTION_NAME"]
    ```
    _config.yml_ **now**
    ```yaml
    opencti:
      ...
    connector:
      ...
    radar:
      feed_lists:
        collection_name_1: id_1
        collection_name_2: id_2
    ```
    See #4194 for more explanations. cc @romain-filigran

Powlinett commented on external-import/socradar/src/lib/config_loader.py:
```diff
@@ -0,0 +1,324 @@
+    log_level: Literal[
+        "debug",
+        "info",
+        "warn",
```
`warn` and `warning` are aliases and the code reflects the doc, do you think it's worth to change the Literal though?

Powlinett replied:
I can't remove `warn` (breaking change) so I added `warning` too

Powlinett commented on external-import/socradar/src/lib/api_client.py:
```diff
@@ -0,0 +1,106 @@
+    """Custom wrapper for exceptions raised in RadarAPIClient"""
+
+
+class RadarFeedItemExtraInfo:
```
It's a good improvement idea indeed but TBH I don't think it's worth it for this connector. The feeds data seems to be really consistent and we have really few issues with this connector so validation seemed a bit overkill for this PR. WDYT?

Powlinett commented on external-import/socradar/src/lib/radar.py:
```diff
@@ -1,364 +1,214 @@
-                        )
-                        stix_batch = []
+                    if len(stix_batch) >= BATCH_MAX_SIZE:
+                        self._handle_batch(stix_batch)
```
The `>=` is a business logic condition, while the `ValueError` error in `_handle_batch` is more an "extra security" for future devs. Maybe it's not useful idk 🤷‍♀️

Powlinett replied:
Yes, it only works if `stix_batch` == `BATCH_MAX_SIZE`.
The `ValueError` is an extra security in the case we delete `if len(stix_batch) >= BATCH_MAX_SIZE` condition before calling  `self._handle_batch` inadvertently. But in fact it should _never be raised_ as it would be the result of a programming error

Powlinett commented on external-import/socradar/.env.sample:
```diff
@@ -0,0 +1,15 @@
+# Connector's custom execution parameters
+RADAR_BASE_FEED_URL=https://platform.socradar.com/api/threat/intelligence/feed_list/ # required
+RADAR_SOCRADAR_KEY=CHANGEME
+RADAR_FEED_LISTS=CHANGEME
```
I fixed _config.yml.sample_ thanks 👍

For the env var, see _README_ here: https://github.com/OpenCTI-Platform/connectors/pull/4196/files#diff-3b5a6003556b2e35ccef9fe7aaafda75b81cf9cee49929006a878cc12542293bR78

pdamoune replied:
The feedlists name seems to be only for logging.

If we only need the IDs, and you deprecated the original configuration, what do you think about keeping a simple list of ids and drop the json format ?

pdamoune commented on external-import/socradar/src/lib/config_loader.py:
```diff
@@ -0,0 +1,324 @@
+    Base class for frozen config models, i.e. not alter-able after `model_post_init()`.
+    """
+
+    model_config = ConfigDict(frozen=True)
```
In order to have the right messages, we need to add  `default_factory`  to the definition of the class

```python
opencti: OpenCTIConfig = Field(default_factory=OpenCTIConfig)
```

- Example of the errors without this parameter :

<img width="847" height="259" alt="image" src="https://github.com/user-attachments/assets/9f222d02-c40f-4c7c-ab85-a2d4c50a8f3b" />
    
- With the `default_factory=OpenCTIConfig`:
    
<img width="847" height="259" alt="image" src="https://github.com/user-attachments/assets/88caea14-a7f7-4d56-ad27-49c8e80e6161" />

Powlinett replied:
IMO the two errors mean two different things:
1. the config file has not been found or the config file doesn't contain any of `opencti`/`connector`/`socprime` keys
2. the config file has been found, it contains all of `opencti`/`connector`/`socprime` keys but the required keys `url` and `token` are missing in `opencti` object

pdamoune commented on external-import/socradar/src/lib/radar.py:
```diff
@@ -1,364 +1,214 @@
-                    if new_objs:
-                        stix_batch.extend(new_objs)
+                for stix_objects in self._convert_feed_items(feed_items):
+                    stix_batch.extend(stix_objects)
```
Nope, just about duplicated data, however it's handled and idempotent in the workflow, so if it's not relevant no worries

pdamoune commented on external-import/socradar/src/lib/api_client.py:
```diff
@@ -0,0 +1,106 @@
+            if first_seen_date
+            else None
+        )
+        self.latest_seen_date: datetime = (
```
same here

pdamoune commented on external-import/socradar/src/lib/api_client.py:
```diff
@@ -0,0 +1,106 @@
+            else None
+        )
+        self.latest_seen_date: datetime | None = (
+            datetime.fromisoformat(latest_seen_date).astimezone(tzinfo=timezone.utc)
```
```suggestion
            datetime.fromisoformat(latest_seen_date).astimezone(tz=timezone.utc)
```

pdamoune reviewed (changes requested):
Thanks @Powlinett !

I request changes on my comment, also, is there a reason why we had only 1 work of ~2k entities before and now 6 or 7 of ~300 entities ?

Otherwise everything seems cool to me

Powlinett commented:
Hi @pdamoune ,

- `confidence_level` has been removed on purpose, as it doesn't meet our new guidelines. Here you can find a conversation explaining the differences between setting `CONNECTOR_CONFIDENCE_LEVEL`(deprecated) and setting the confidence level on the user impersonating a connector: https://github.com/OpenCTI-Platform/connectors/pull/3316#issuecomment-2654371084

- the original creation date is an unexpected change, I fixed that, thanks 👍

- the entities per work count is an unexpected changes too, also fixed! 🙏

pdamoune commented on external-import/socradar/src/lib/config_loader.py:
```diff
@@ -0,0 +1,286 @@
+        connector_data: dict = data.get("connector", {})
+        radar_data: dict = data.get("radar", {})
+
+        if interval := radar_data.pop("run_interval", None):
```
This only work for the environment variables, but will break the config.yml variables.
if we want to keep the compatibility we d need to check on `radar_run_interval` too

Powlinett replied:
Same answer as above ( + PR's description)

Powlinett commented on external-import/socradar/src/config.yml.sample:
```diff
@@ -4,21 +4,14 @@ opencti:
+  duration_period: 'PT10M'
 
 radar:
-  radar_base_feed_url: "https://platform.socradar.com/api/threat/intelligence/feed_list/"
-  radar_socradar_key: "SOCRADAR_KEY"
```
I assumed breaking changes in **development mode only** in order to keep the `ConfigLoader` as concise as possible. See PR's description.
